### PR TITLE
RavenDB-18353 Alphanumerical sorting - server side

### PR DESCRIPTION
**Pull request by maciejaszyk**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18353

### Additional description

Integration of comparer.

### Type of change

- New feature

### How risky is the change?

- Not relevant

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Corax/Queries/IMatchComparer.cs [hunk 1]
```diff
@@ -3,12 +3,19 @@
 
 namespace Corax.Queries
 {
+    public enum SortingType
```
Comment (ayende) [inline src/Corax/Queries/IMatchComparer.cs:6]: why is this needed?

File: src/Corax/Queries/SortingMatch.Comparers.cs [hunk 2]
```diff
@@ -20,7 +15,158 @@ internal static int CompareAscending(ReadOnlySpan<byte> x, ReadOnlySpan<byte> y)
...
+            private static byte ToUpper(byte c)
+            {
+                if (c >= 'a' && c <= 'z')
+                    return (byte)(c - ('a' - 'A'));
+
+                return c; 
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static int CompareAlphanumericAscending(ReadOnlySpan<byte> a, ReadOnlySpan<byte> b, bool foldCase = false)
```
Comment (ayende) [inline src/Corax/Queries/SortingMatch.Comparers.cs]: This is valid only to ASCII buffers

Comment (maciejaszyk) [reply src/Corax/Queries/SortingMatch.Comparers.cs]: @redknightlois

File: src/Corax/Queries/IMatchComparer.cs (hunk 1 above)
Comment (maciejaszyk) [reply src/Corax/Queries/IMatchComparer.cs:6]: This is a workaround for the current Corax API.

@redknightlois can you add `Alphanumeric` to `MatchCompareFieldType`? Because alphanumeric will always operate on strings, I think.

File: src/Corax/Queries/SortingMatch.Comparers.cs (hunk 2 above)
Comment (redknightlois) [reply src/Corax/Queries/SortingMatch.Comparers.cs]: Yes, we need to check the buffer for ASCIIness and go back to standard comparer. Unless you know how to do that with Unicode (UTF-8).

Comment (ayende) [reply src/Corax/Queries/SortingMatch.Comparers.cs]: Alpha numeric is rare enough that we don't need to specialize here. Can always use the usual comparere.

File: src/Corax/Queries/IMatchComparer.cs (hunk 1 above)
Comment (redknightlois) [reply src/Corax/Queries/IMatchComparer.cs:6]: Natural Order is only reasonable to work when you are doing roman language encodings (ASCII). I dont think we need to include anything else, this is kinda a replacement of the normal sorting only on certain occasions (should be Opt-in) but that is something that only RavenDB can know.

**Comment by redknightlois** [comment]
Support for UTF8: https://github.com/redknightlois/ravendb/commit/16afc7c6d6325f73c2655d063fd95886a6831e99
cc @maciejaszyk

**Comment by redknightlois** [comment]
Commit for fixed handling of capital letters: 6ed9f52c8d03e4d80364afcf2c39fe4fbad35eba
Still needed to update the tests to expect differences with Lucene.

**Review by arekpalinski — APPROVED** [review]
